sw/device/lib/dif/dif_i2c.c

```diff
@@ -0,0 +1,661 @@
+                                              .index = I2C_TIMING1_T_F_OFFSET,
+                                              .value = timing.fall_time,
+                                          });
+  mmio_region_write32(i2c->base_addr, I2C_TIMING1_REG_OFFSET, timing1);
+
+  uint32_t timing2 = 0;
+  timing2 =
+      bitfield_set_field32(timing2, (bitfield_field32_t){
+                                        .mask = I2C_TIMING2_TSU_STA_MASK,
+                                        .index = I2C_TIMING2_TSU_STA_OFFSET,
+                                        .value = timing.start_signal_setup_time,
+                                    });
+  timing2 =
+      bitfield_set_field32(timing2, (bitfield_field32_t){
+                                        .mask = I2C_TIMING2_THD_STA_MASK,
+                                        .index = I2C_TIMING2_THD_STA_OFFSET,
+                                        .value = timing.start_signal_hold_time,
+                                    });
+  mmio_region_write32(i2c->base_addr, I2C_TIMING2_REG_OFFSET, timing2);
+
+  uint32_t timing3 = 0;
+  timing3 =
+      bitfield_set_field32(timing3, (bitfield_field32_t){
+                                        .mask = I2C_TIMING3_TSU_DAT_MASK,
+                                        .index = I2C_TIMING3_TSU_DAT_OFFSET,
+                                        .value = timing.data_signal_setup_time,
+                                    });
+  timing3 =
+      bitfield_set_field32(timing3, (bitfield_field32_t){
+                                        .mask = I2C_TIMING3_THD_DAT_MASK,
+                                        .index = I2C_TIMING3_THD_DAT_OFFSET,
+                                        .value = timing.data_signal_hold_time,
+                                    });
+  mmio_region_write32(i2c->base_addr, I2C_TIMING3_REG_OFFSET, timing3);
+
+  uint32_t timing4 = 0;
+  timing4 =
+      bitfield_set_field32(timing4, (bitfield_field32_t){
+                                        .mask = I2C_TIMING4_TSU_STO_MASK,
+                                        .index = I2C_TIMING4_TSU_STO_OFFSET,
+                                        .value = timing.stop_signal_setup_time,
+                                    });
+  timing4 =
+      bitfield_set_field32(timing4, (bitfield_field32_t){
+                                        .mask = I2C_TIMING4_T_BUF_MASK,
+                                        .index = I2C_TIMING4_T_BUF_OFFSET,
+                                        .value = timing.stop_signal_hold_time,
+                                    });
+  mmio_region_write32(i2c->base_addr, I2C_TIMING4_REG_OFFSET, timing4);
+}
+
+dif_i2c_result_t dif_i2c_init(mmio_region_t base_addr,
+                              dif_i2c_timing_params_t timing, dif_i2c_t *i2c) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  i2c->base_addr = base_addr;
+
+  // Turn the device off, before resetting it.
+  (void)dif_i2c_host_set_enabled(i2c, kDifI2cDisabled);
```

Comment:
Handle error returns. I know that you're calling with a constant and right now it doesn't have any different error path, but it's defensive to check.

Reply:
Meh, ok.